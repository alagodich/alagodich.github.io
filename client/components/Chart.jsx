import React, {Component} from 'react';
import PropTypes from 'prop-types';
import Chord from './Chord.jsx';

const propTypes = {
    data: PropTypes.object
};

class Chart extends Component {

    renderChart() {
        const items = [];
        let row = [],
            counter = 0,
            inRepeat = false;

        function processBarline(barlines) {
            let left, right, top;
            if (!barlines) {
                return [null, null, null];
            }
            for (const barline of barlines) {
                const style = barline['bar-style'];

                if (barline.$.location === 'left' && style) {
                    left = `barline--left--${style}`;
                }
                if (barline.$.location === 'right' && style) {
                    right = `barline--right--${style}`;
                }

                if (barline.$.location === 'right' && barline.ending && barline.ending[0].$.type === 'discontinue') {
                    top = 'barline--top--light';
                }
            }
            return [left, right, top];
        }

        function getBarlinesClassnames(barlines, first) {
            const classNames = [],
                barlineClassNames = processBarline(barlines),
                top = barlineClassNames[2];
            let left = barlineClassNames[0],
                right = barlineClassNames[1];

            if (!left && first) {
                left = 'barline--left--light';
            }
            if (left) {
                classNames.push(left);
            }

            right = right ? right : 'barline--right--light';
            classNames.push(right);

            if (top) {
                classNames.push(top);
            }

            return classNames;
        }

        function processBarChords(bar) {
            const chords = [];

            bar.harmony.forEach((chord, number) => {
                const duration = bar.note[number].type[0];
                chords.push(<Chord key={number} config={chord} duration={duration}/>);
            });
            return <div className="ui grid chords">{chords}</div>;
        }

        function secondRepeat(bar) {
            if (bar.barline && bar.barline[1] && bar.barline[1].$.location === 'left') {
                if (bar.barline[1].ending[0].$.type === 'start' && bar.barline[1].ending[0].$.number === '2') {
                    return true;
                }
            }
            return false;
        }

        for (const barConfig of this.props.data.measure) {
            const bar = processBarChords(barConfig),
                index = this.props.data.measure.indexOf(barConfig),
                first = counter === 0,
                last = index === this.props.data.measure.length - 1,
                lastInRow = counter === 3,
                repeatEnded = barConfig.barline
                    && barConfig.barline[0]
                    && barConfig.barline[0].$.location === 'right'
                    && barConfig.barline[0]['bar-style']
                    && barConfig.barline[0]['bar-style'][0] === 'light-light';
            let classNames = ['column', 'bar'].concat(getBarlinesClassnames(barConfig.barline, first));

            if (secondRepeat(barConfig)) {
                classNames = classNames.concat(['right', 'floated']);
                inRepeat = true;
            }

            row.push(<div key={index} className={classNames.join(' ')}>{bar}</div>);
            counter++;

            if (last || lastInRow || (inRepeat && repeatEnded)) {
                items.push(<div key={index} className="row">{row}</div>);
                row = [];
                counter = 0;
                inRepeat = false;
            }

        }
        return <div className="ui four column grid chart">{items}</div>;
    }

    render() {
        const chart = this.props.data ? this.renderChart() : 'Loading...';
        return (
            <div>{chart}</div>
        );
    }
}

Chart.propTypes = propTypes;

export default Chart;
