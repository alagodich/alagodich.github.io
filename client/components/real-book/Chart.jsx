import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {Table} from 'semantic-ui-react';
import Chord from './Chord.jsx';

const propTypes = {
    model: PropTypes.shape({
        title: PropTypes.string,
        author: PropTypes.string,
        style: PropTypes.string,
        key: PropTypes.string,
        chordString: PropTypes.string,
        segments: PropTypes.arrayOf(PropTypes.shape({
            name: PropTypes.string,
            data: PropTypes.arrayOf(PropTypes.shape({
                chords: PropTypes.string,
                openingLine: PropTypes.string,
                closingLine: PropTypes.string,
                ending: PropTypes.string,
                timeSignature: PropTypes.string,
                divider: PropTypes.string,
                coda: PropTypes.bool
            }))
        }))
    })
};

const styles = {
    table: {
        fontWeight: 500,
        fontSize: 20
    },
    sectionHeader: {
        fontWeight: 900,
        fontSize: 35
    }
};

class Chart extends PureComponent {
    processLines(segment) {
        if (!segment) {
            return null;
        }
        segment.lines = [];

        let line = [];

        segment.data.forEach(bar => {
            // Not rendering dividers for now
            if (bar.divider) {
                return;
            }

            if (line.length < 4) {
                // If it is a last bar in line, and has no closing line, add default
                if (line.length === 3 && !bar.closingLine) {
                    bar.closingLine = '|';
                }

                line.push(bar);

                // If current closing bar line is not regular and current line is not and ending line, break the line
                if (bar.closingLine && bar.closingLine !== '|' && line[0] && !line[0].ending) {
                    segment.lines.push(line);
                    line = [];
                }
            } else {
                segment.lines.push(line);
                line = [bar];
            }
        });
        if (line.length) {
            // If the last line is less than 4 bars and it is an ending line, fill it to the size of 4
            if (line.length < 4 && line[0].ending) {
                const filler = new Array(4 - line.length).fill({empty: true});

                line = [...filler, ...line];
            }
            segment.lines.push(line);
        }

        return segment;
    }

    renderChart() {
        const tableRows = [];
        const segments = this.props.model.segments
            .map(segment => this.processLines(segment));

        segments.forEach((segment, segmentKey) => {
            const headerCell = (
                <Table.Cell
                    width={1}
                    verticalAlign="top"
                    style={styles.sectionHeader}
                    className="chart__bar"
                >
                    {segment.name}
                </Table.Cell>
            );

            segment.lines.forEach((line, key) => {
                tableRows.push(
                    <Table.Row key={`${segmentKey}-${key}`} className="chart__bar-line">
                        {key === 0 ? headerCell : <Table.Cell className="chart__bar" width={1} />}
                        {line.map((bar, key) => <Chord key={key} {...bar} />)}
                    </Table.Row>
                );
            });
        });

        return (
            <Table
                basic="very"
                singleLine
                fixed
                columns={5}
                style={styles.table}
                className="chart"
                unstackable
                attached="top"
            >
                <Table.Body>
                    {tableRows}
                </Table.Body>
            </Table>
        );
    }

    render() {
        return <div>{this.props.model ? this.renderChart() : 'Loading...'}</div>;
    }
}

Chart.propTypes = propTypes;

export default Chart;
