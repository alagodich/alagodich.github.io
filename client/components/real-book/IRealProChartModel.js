import {timeSignatures, endings} from './IRealProUrlParser';
import {default as _omit} from 'lodash/omit';

export default class IRealProChartModel {
    constructor(props) {
        if (!props) {
            return;
        }

        this.title = props.title;
        this.author = props.author;
        this.style = props.style;
        this.key = props.key;
        this.chordString = props.chordString;

        this.init();
    }

    /**
     * Prepare chart object from chart string
     *
     * Split chart by named sections (A, B, C)
     */
    init() {
        const segments = this.chordString
            /**
             * In some cases segment name *A, *B may be inside bar lines, in some outside,
             * we will put them always outside
             */
            .replace(/([{[|Y])(\*\w)/g, '$2$1')
            // Split chord string by Segment name *A, *B, they are all now outside the bar lines
            .match(/(\*\w)([^*]+)/g);

        if (!segments) {
            this.segments = [
                {name: '', data: this.parseSegment(this.chordString)}
            ];
        } else {
            this.segments = segments
                .map(segmentString => ({
                    name: segmentString[1],
                    data: this.parseSegment(segmentString.substring(2))
                }));
        }
    }

    /**
     * Split data string by segments
     * @param segmentString
     * @returns {*}
     */
    parseSegment(segmentString) {
        const segment = [];

        // Process segment data string to form final chord collection
        segmentString
            // Split string by bar line separator
            .match(/([{}[\]|ZY])([^{}[\]|ZY]*)/g)
            // Process bar collection
            .forEach(rawBarDataString => {
                let barString = rawBarDataString;

                // Process repeat sign (r)
                if (barString.includes('r') && barString !== 'r') {
                    // Split measure with (r), cut r from it and process the remains as usual
                    const rSplitMatch = barString.match(/([^r]*)(r)([^r]*)/);
                    const last2Bars = segment.slice(-2).map(bar => {
                        const newBar = Object.assign({}, bar);

                        // Cloned bar will always have simple opening bar line
                        newBar.openingLine = '|';
                        return _omit(newBar, 'timeSignature');
                    });

                    // If last of the cloned bar has special closing bar line, keep it only on cloned pair
                    if (last2Bars[1].closingLine && last2Bars[1].closingLine !== '|') {
                        segment[segment.length - 1].closingLine = '|';
                    }
                    segment.push(...last2Bars);
                    // Remaining measure string with (r) sign
                    barString = [rSplitMatch[1], rSplitMatch[3]].join('');
                }

                const bar = this.parseBar(barString);

                // Merge some parts together, for example closing bar line with the previous bar
                if (bar.closingLine && Object.getOwnPropertyNames(bar).length === 1) {
                    if (!segment[segment.length - 1]) {
                        throw new Error(`Invalid segment, closing bar as a first part: ${segmentString}`);
                    }
                    segment[segment.length - 1].closingLine = bar.closingLine;
                } else {
                    segment.push(bar);
                }
            });

        if (!segment[segment.length - 1].closingLine) {
            segment[segment.length - 1].closingLine = ']';
        }

        return segment;
    }

    /**
     * Parse bar string, find all possible elements and return bar props object
     *
     * @param barString
     * @returns {{}}
     */
    // eslint-disable-next-line complexity,max-statements
    parseBar(barString) {
        let rawBarString = barString.trim();
        const barProps = {};

        // Check and extract time signature
        const timeSignature = rawBarString.match(/(T\d\d)/g);

        if (timeSignature && timeSignature[0]) {
            if (!timeSignatures[timeSignature[0]]) {
                // console.log('Invalid time signature', timeSignature);
                throw new Error(`Invalid time signature ${rawBarString}`);
            }
            barProps.timeSignature = timeSignatures[timeSignature[0]];
            rawBarString = rawBarString.split(timeSignature[0]).join('');
        }

        // Check and extract ending
        const ending = rawBarString.match(/(N\d)/g);

        if (ending && ending[0]) {
            if (!endings.includes(ending[0])) {
                throw new Error(`Invalid ending ${rawBarString}`);
            }
            barProps.ending = ending[0];
            rawBarString = rawBarString.split(ending[0]).join('');
        }

        // Check and extract coda
        const hasCoda = rawBarString.includes('Q');

        if (hasCoda) {
            barProps.coda = true;
            rawBarString = rawBarString.split('Q').join('');
        }

        // Check if that part is a divider
        const isDivider = rawBarString === 'Y';

        if (isDivider) {
            barProps.divider = 'Y';
            rawBarString = '';
        }

        // Handle closing brackets
        if (rawBarString.length === 1) {
            const closingBarLine = rawBarString.match(/([{}[\]|Z])/g);

            if (closingBarLine.length !== 1) {
                throw new Error(`Unexpected rawBarString ${rawBarString}`);
            }
            barProps.closingLine = closingBarLine[0];
            rawBarString = rawBarString.split(closingBarLine[0]).join('');
        }

        // Find the final opening line and chords
        const barParts = rawBarString.match(/([{}[\]|Z]+)([^{}[\]|Z]*)/);

        if (barParts && barParts[0] && barParts[1] && barParts[2]) {
            barProps.openingLine = barParts[1];
            barProps.chords = barParts[2].trim();

            rawBarString = rawBarString.split(barParts[0]).join('');
        }

        // After opening line and a chord there should be nothing more left
        if (rawBarString.length > 0) {
            throw new Error(`Unparsed rawBarString ${rawBarString}`);
        }

        return barProps;
    }

}
