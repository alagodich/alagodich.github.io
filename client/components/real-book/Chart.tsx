import React, {ReactElement, useEffect} from 'react';
import {Segment, Grid} from 'semantic-ui-react';
// import {ChartBar} from './ChartBar';
import {SvgChartBar} from './SvgChartBar';
import {RouteComponentProps, Link, useHistory} from 'react-router-dom';
import {useDispatch, useSelector} from 'react-redux';
import {loadSong} from './store/library-slice';
import IRealProChartModel from './IRealProChartModel';
import {IIRealProChartBar, IIRealProChartSegment} from './types';
import {RootState} from './store/reducer';

interface IChartProps {
    playlist: string;
    songId: string;
}

/**
 * If the ending line is less than 4 bars, fill it to the size of 4
 *
 * @param line
 * @param lineLength
 */
function fillEndingLineToSize4WithEmptyBars(line: IIRealProChartBar[], lineLength = 4) {
    if (line.length < lineLength && line[0].ending) {
        const filler = new Array(lineLength - line.length).fill({empty: true});

        return [...filler, ...line];
    }

    return line;
}

export function processLines(segment: IIRealProChartSegment): IIRealProChartBar[][] {
    if (!segment || !segment.data) {
        return [];
    }
    const lines: IIRealProChartBar[][] = [];
    let line: IIRealProChartBar[] = [];

    // eslint-disable-next-line complexity
    segment.data.forEach(barData => {
        const bar = Object.assign({}, barData);

        // Not rendering dividers for now, but it should break the line
        if (bar.divider) {
            if (line.length) {
                lines.push(fillEndingLineToSize4WithEmptyBars(line, lines[lines.length - 1]?.length));
                line = [];
            }
            return;
        }

        if (line.length < 4) {
            // If it is a last bar in line, and has no closing line, add default
            if (line.length === 3 && !bar.close) {
                bar.close = '|';
            }

            // If current closing bar line is not regular and current line is not and ending line, break the line
            if (bar.close && bar.close !== '|' && line[0] && !line[0].ending) {
                // Fix closing bar lines, if it has opening sign for some reason
                if (bar.close && ['[', '}'].includes(bar.close)) {
                    bar.close = bar.close === '['
                        ? ']'
                        : '}';
                }
                line.push(bar);
                lines.push(line);
                line = [];
            } else {
                line.push(bar);
            }
        } else {
            lines.push(line);
            line = [bar];
        }
    });

    if (line.length) {
        // If the last line is less than 4 bars and it is an ending line, fill it to the size of 4
        lines.push(fillEndingLineToSize4WithEmptyBars(line, lines[lines.length - 1]?.length));
    }

    return lines;
}

export const Chart = React.memo((props: RouteComponentProps<IChartProps>): ReactElement | null => {
    const dispatch = useDispatch();
    const history = useHistory();

    const {playlist, songId} = props.match.params;
    const {
        songs,
        activePlaylist,
        activeSong,
        error,
        displayType
    } = useSelector((state: RootState) => state.library);
    const {notation} = useSelector((state: RootState) => state.chart);

    useEffect(() => {
        dispatch(loadSong(playlist, parseInt(songId, 10)));
    }, [playlist, songId]);

    useEffect(() => {
        // eslint-disable-next-line complexity
        function handleKeyPress(event: React.KeyboardEvent): void {
            if ((!activeSong && activeSong !== 0) || displayType !== 'chart') {
                return;
            }

            if (event.key === 'Escape' || event.key === 'q') {
                history.push(`/${activePlaylist}`);
            } else if (event.key === 'ArrowLeft') {
                if (songs[activeSong - 1]) {
                    history.push(`/${activePlaylist}/${activeSong - 1}`);
                }
            } else if (event.key === 'ArrowRight') {
                if (songs[activeSong + 1]) {
                    history.push(`/${activePlaylist}/${activeSong + 1}`);
                }
            }
        }

        document.addEventListener('keyup', handleKeyPress as any, false);
        return function cleanUp() {
            document.removeEventListener('keyup', handleKeyPress as any, false);
        };
    });

    function renderChart(): ReactElement | null {
        const segments: ReactElement[] = [];

        if (!activeSong && activeSong !== 0) {
            return null;
        }
        const shouldSuppressTuneAdjective = notation === 'berklee';
        const model = new IRealProChartModel(songs[activeSong], shouldSuppressTuneAdjective);

        model.segments.forEach((segment, segmentKey) => {
            const lines = processLines(segment);
            const gridRows: ReactElement[] = [];

            if (!lines) {
                return;
            }

            lines.forEach((line: IIRealProChartBar[], key) => {
                gridRows.push(
                    <Grid.Row key={key}>
                        {line.map((bar, barKey) => (
                            <Grid.Column width={4} key={barKey}>
                                <SvgChartBar {...bar} notation={notation} />
                            </Grid.Column>
                        ))}
                    </Grid.Row>
                );
            });

            segments.push(
                <Grid columns={16} className="chart" key={segmentKey}>
                    <Grid.Column width={1} className="chart__section-header">
                        {segment.name ?? ' '}
                    </Grid.Column>
                    <Grid.Column width={15}>
                        <Grid columns={4}>
                            {gridRows}
                        </Grid>
                    </Grid.Column>
                </Grid>
            );
        });

        return (
            <>
                {model.errors.length
                    ? model.errors.map((modelError: string, key) => (
                        <Segment basic inverted key={key}>{`Error: ${modelError}`}</Segment>
                    ))
                    : null
                }
                {segments}
            </>
        );
    }

    return error
        ? <div>{error} <Link to={`/${playlist}`}>{'Back to playlist'}</Link></div>
        : (activeSong || activeSong === 0) ? renderChart() : null;
});

export default Chart;
