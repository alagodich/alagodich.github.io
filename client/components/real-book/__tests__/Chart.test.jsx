/* eslint max-len: 0 */

import Chart from '../Chart.jsx';
import React from 'react';
import {create} from 'react-test-renderer';
import IRealProChartModel from '../IRealProChartModel';

describe('Chart Component', () => {
    it('should render empty', () => {
        const element = create(<Chart/>);
        const tree = element.toJSON();

        expect(tree).toMatchSnapshot();
    });
    it('should render Afternoon in Paris', () => {
        const props = {
            title: 'Afternoon In Paris',
            author: 'Lewis John',
            style: 'Medium Swing',
            key: 'C',
            chordString: '*A{T44C^7 |C-7 F7|Bb^7 |Bb-7 Eb7|Ab^7 |D-7 G7#9|N1C^7 |D-7 G7 } |N2C^7 |x ]*B[D-7 |G7 |C^7/E |A7 |Y|D-7 |G7 |C^7(C#-7) (F#7)|D-7 G7 ]*A[C^7 |C-7 F7|Bb^7 |Bb-7 Eb7|Ab^7 |D-7 G7#9|C^7 |D-7 G7 Z'
        };
        const model = new IRealProChartModel(props);

        const element = create(<Chart model={model} />);
        const tree = element.toJSON();

        expect(tree).toMatchSnapshot();
    });
    it('should render 500 Miles High', () => {
        const props = {
            title: '500 Miles High',
            author: 'Corea Chick',
            style: 'Bossa Nova',
            key: 'E-',
            chordString: '[T44E-7 |x |G-7 |x |Bb^7 |x |Bh7 |E7#9 |A-7 |x |F#h7 |x |F-7 |x Q|C-7 |x |B7#9 |x Z Y{QC-7 |x |Ab^7 |x }'
        };
        const model = new IRealProChartModel(props);

        const element = create(<Chart model={model} />);
        const tree = element.toJSON();

        expect(tree).toMatchSnapshot();
    });
    describe('processLines', () => {
        it('should handle empty call', () => {
            const element = create(<Chart/>);
            const instance = element.getInstance();

            expect(instance.processLines()).toBeNull();
            expect(instance.processLines('')).toBeNull();
            expect(instance.processLines(null)).toBeNull();
        });
        it('should correctly process segment with repeat', () => {
            const element = create(<Chart/>);
            const instance = element.getInstance();
            const segment = {
                name: 'A',
                data: [
                    {timeSignature: '4 / 4', openingLine: '{', chords: 'C^7'},
                    {openingLine: '|', chords: 'C-7 F7'},
                    {openingLine: '|', chords: 'Bb^7'},
                    {openingLine: '|', chords: 'Bb-7 Eb7'},
                    {openingLine: '|', chords: 'Ab^7'},
                    {openingLine: '|', chords: 'D-7 G7#9'},
                    {ending: 'N1', openingLine: '|', chords: 'C^7'},
                    {openingLine: '|', chords: 'D-7 G7', closingLine: '}'},
                    {ending: 'N2', openingLine: '|', chords: 'C^7'},
                    {openingLine: '|', chords: 'x', closingLine: ']'}
                ]
            };

            expect(instance.processLines(segment)).toEqual({
                name: 'A',
                data: [
                    {timeSignature: '4 / 4', openingLine: '{', chords: 'C^7'},
                    {openingLine: '|', chords: 'C-7 F7'},
                    {openingLine: '|', chords: 'Bb^7'},
                    {openingLine: '|', chords: 'Bb-7 Eb7', closingLine: '|'},
                    {openingLine: '|', chords: 'Ab^7'},
                    {openingLine: '|', chords: 'D-7 G7#9'},
                    {ending: 'N1', openingLine: '|', chords: 'C^7'},
                    {openingLine: '|', chords: 'D-7 G7', closingLine: '}'},
                    {ending: 'N2', openingLine: '|', chords: 'C^7'},
                    {openingLine: '|', chords: 'x', closingLine: ']'}
                ],
                lines: [
                    [
                        {timeSignature: '4 / 4', openingLine: '{', chords: 'C^7'},
                        {openingLine: '|', chords: 'C-7 F7'},
                        {openingLine: '|', chords: 'Bb^7'},
                        {openingLine: '|', chords: 'Bb-7 Eb7', closingLine: '|'}
                    ],
                    [
                        {openingLine: '|', chords: 'Ab^7'},
                        {openingLine: '|', chords: 'D-7 G7#9'},
                        {ending: 'N1', openingLine: '|', chords: 'C^7'},
                        {openingLine: '|', chords: 'D-7 G7', closingLine: '}'}
                    ],
                    [
                        {empty: true},
                        {empty: true},
                        {ending: 'N2', openingLine: '|', chords: 'C^7'},
                        {openingLine: '|', chords: 'x', closingLine: ']'}
                    ]
                ]
            });
        });

        it('should process not /4 number of chords with no repeat', () => {
            const element = create(<Chart/>);
            const instance = element.getInstance();
            const segment = {
                name: '',
                data: [
                    {timeSignature: '4 / 4', openingLine: '[', chords: 'E-7'},
                    {openingLine: '|', chords: 'x'},
                    {openingLine: '|', chords: 'G-7'},
                    {openingLine: '|', chords: 'x'},
                    {openingLine: '|', chords: 'F-7'},
                    {coda: true, openingLine: '|', chords: 'x'},
                    {openingLine: '|', chords: 'C-7'},
                    {openingLine: '|', chords: 'x'},
                    {openingLine: '|', chords: 'B7#9'},
                    {openingLine: '|', chords: 'x', closingLine: 'Z'},
                    {divider: 'Y'},
                    {coda: true, openingLine: '{', chords: 'C-7'},
                    {openingLine: '|', chords: 'x'},
                    {openingLine: '|', chords: 'Ab^7'},
                    {openingLine: '|', chords: 'x', closingLine: '}'}
                ]
            };

            expect(instance.processLines(segment)).toEqual({
                name: '',
                data: [
                    {timeSignature: '4 / 4', openingLine: '[', chords: 'E-7'},
                    {openingLine: '|', chords: 'x'},
                    {openingLine: '|', chords: 'G-7'},
                    {openingLine: '|', chords: 'x', closingLine: '|'},

                    {openingLine: '|', chords: 'F-7'},
                    {coda: true, openingLine: '|', chords: 'x'},
                    {openingLine: '|', chords: 'C-7'},
                    {openingLine: '|', chords: 'x', closingLine: '|'},

                    {openingLine: '|', chords: 'B7#9'},
                    // Closing bar line other than | should change the line
                    {openingLine: '|', chords: 'x', closingLine: 'Z'},
                    // Ignoring divider for now
                    {divider: 'Y'},

                    {coda: true, openingLine: '{', chords: 'C-7'},
                    {openingLine: '|', chords: 'x'},
                    {openingLine: '|', chords: 'Ab^7'},
                    {openingLine: '|', chords: 'x', closingLine: '}'}
                ],
                lines: [
                    [
                        {timeSignature: '4 / 4', openingLine: '[', chords: 'E-7'},
                        {openingLine: '|', chords: 'x'},
                        {openingLine: '|', chords: 'G-7'},
                        {openingLine: '|', chords: 'x', closingLine: '|'}
                    ],
                    [
                        {openingLine: '|', chords: 'F-7'},
                        {coda: true, openingLine: '|', chords: 'x'},
                        {openingLine: '|', chords: 'C-7'},
                        {openingLine: '|', chords: 'x', closingLine: '|'}
                    ],
                    [
                        {openingLine: '|', chords: 'B7#9'},
                        {openingLine: '|', chords: 'x', closingLine: 'Z'}
                    ],
                    [
                        {coda: true, openingLine: '{', chords: 'C-7'},
                        {openingLine: '|', chords: 'x'},
                        {openingLine: '|', chords: 'Ab^7'},
                        {openingLine: '|', chords: 'x', closingLine: '}'}
                    ]
                ]
            });
        });
    });
});
