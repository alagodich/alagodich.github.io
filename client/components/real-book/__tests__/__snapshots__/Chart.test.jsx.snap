// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart Component should render 500 Miles High 1`] = `
<div>
  <table
    className="ui fixed single line unstackable top attached very basic five column table chart"
    style={
      Object {
        "fontSize": 20,
        "fontWeight": 500,
      }
    }
  >
    <tbody
      className=""
    >
      <tr
        className="chart__bar-line"
      >
        <td
          className="top aligned one wide chart__bar"
          style={
            Object {
              "fontSize": 35,
              "fontWeight": 900,
            }
          }
        >
          
        </td>
        <td
          className="four wide chart__bar barline--left--light-light"
        >
          <span>
            E-7
          </span>
        </td>
        <td
          className="center aligned four wide chart__bar barline--left--light"
        >
          <span
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            %
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            G-7
          </span>
        </td>
        <td
          className="center aligned four wide chart__bar barline--left--light barline--right--light"
        >
          <span
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            %
          </span>
        </td>
      </tr>
      <tr
        className="chart__bar-line"
      >
        <td
          className="one wide chart__bar"
        />
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            Bb^7
          </span>
        </td>
        <td
          className="center aligned four wide chart__bar barline--left--light"
        >
          <span
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            %
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            Bh7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light barline--right--light"
        >
          <span>
            E7#9
          </span>
        </td>
      </tr>
      <tr
        className="chart__bar-line"
      >
        <td
          className="one wide chart__bar"
        />
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            A-7
          </span>
        </td>
        <td
          className="center aligned four wide chart__bar barline--left--light"
        >
          <span
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            %
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            F#h7
          </span>
        </td>
        <td
          className="center aligned four wide chart__bar barline--left--light barline--right--light"
        >
          <span
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            %
          </span>
        </td>
      </tr>
      <tr
        className="chart__bar-line"
      >
        <td
          className="one wide chart__bar"
        />
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            F-7
          </span>
        </td>
        <td
          className="center aligned four wide chart__bar barline--left--light"
        >
          <span
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            %
          </span>
          <span
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            (Coda)
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            C-7
          </span>
        </td>
        <td
          className="center aligned four wide chart__bar barline--left--light barline--right--light"
        >
          <span
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            %
          </span>
        </td>
      </tr>
      <tr
        className="chart__bar-line"
      >
        <td
          className="one wide chart__bar"
        />
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            B7#9
          </span>
        </td>
        <td
          className="center aligned four wide chart__bar barline--left--light barline--right--light-heavy"
        >
          <span
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            %
          </span>
        </td>
      </tr>
      <tr
        className="chart__bar-line"
      >
        <td
          className="one wide chart__bar"
        />
        <td
          className="four wide chart__bar barline--left--heavy-dotted"
        >
          <span>
            C-7
          </span>
          <span
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            (Coda)
          </span>
        </td>
        <td
          className="center aligned four wide chart__bar barline--left--light"
        >
          <span
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            %
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            Ab^7
          </span>
        </td>
        <td
          className="center aligned four wide chart__bar barline--left--light barline--right--dotted-heavy"
        >
          <span
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            %
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Chart Component should render Afternoon in Paris 1`] = `
<div>
  <table
    className="ui fixed single line unstackable top attached very basic five column table chart"
    style={
      Object {
        "fontSize": 20,
        "fontWeight": 500,
      }
    }
  >
    <tbody
      className=""
    >
      <tr
        className="chart__bar-line"
      >
        <td
          className="top aligned one wide chart__bar"
          style={
            Object {
              "fontSize": 35,
              "fontWeight": 900,
            }
          }
        >
          A
        </td>
        <td
          className="four wide chart__bar barline--left--heavy-dotted"
        >
          <span>
            C^7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            C-7 F7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            Bb^7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light barline--right--light"
        >
          <span>
            Bb-7 Eb7
          </span>
        </td>
      </tr>
      <tr
        className="chart__bar-line"
      >
        <td
          className="one wide chart__bar"
        />
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            Ab^7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            D-7 G7#9
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light barline--top--light"
        >
          <span>
            C^7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light barline--right--dotted-heavy"
        >
          <span>
            D-7 G7
          </span>
        </td>
      </tr>
      <tr
        className="chart__bar-line"
      >
        <td
          className="one wide chart__bar"
        />
        <td
          className="four wide chart__bar"
        >
          <span />
        </td>
        <td
          className="four wide chart__bar"
        >
          <span />
        </td>
        <td
          className="four wide chart__bar barline--left--light barline--top--light"
        >
          <span>
            C^7
          </span>
        </td>
        <td
          className="center aligned four wide chart__bar barline--left--light barline--right--light-light"
        >
          <span
            style={
              Object {
                "opacity": 0.5,
              }
            }
          >
            %
          </span>
        </td>
      </tr>
      <tr
        className="chart__bar-line"
      >
        <td
          className="top aligned one wide chart__bar"
          style={
            Object {
              "fontSize": 35,
              "fontWeight": 900,
            }
          }
        >
          B
        </td>
        <td
          className="four wide chart__bar barline--left--light-light"
        >
          <span>
            D-7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            G7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            C^7/E
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light barline--right--light"
        >
          <span>
            A7
          </span>
        </td>
      </tr>
      <tr
        className="chart__bar-line"
      >
        <td
          className="one wide chart__bar"
        />
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            D-7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            G7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            C^7(C#-7) (F#7)
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light barline--right--light-light"
        >
          <span>
            D-7 G7
          </span>
        </td>
      </tr>
      <tr
        className="chart__bar-line"
      >
        <td
          className="top aligned one wide chart__bar"
          style={
            Object {
              "fontSize": 35,
              "fontWeight": 900,
            }
          }
        >
          A
        </td>
        <td
          className="four wide chart__bar barline--left--light-light"
        >
          <span>
            C^7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            C-7 F7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            Bb^7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light barline--right--light"
        >
          <span>
            Bb-7 Eb7
          </span>
        </td>
      </tr>
      <tr
        className="chart__bar-line"
      >
        <td
          className="one wide chart__bar"
        />
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            Ab^7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            D-7 G7#9
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light"
        >
          <span>
            C^7
          </span>
        </td>
        <td
          className="four wide chart__bar barline--left--light barline--right--light-heavy"
        >
          <span>
            D-7 G7
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Chart Component should render empty 1`] = `
<div>
  Loading...
</div>
`;
